### DISCUSSION TITLE
Onnx padim output processing in C#

### DISCUSSION THREAD

**Discussion by Tibe31** [open-edge-platform/anomalib#1539] [Q&A] [answered]
### What is the motivation for this task?

I've trained a padim anomaly detector on my own dataset; now I want to deploy it in C#.
I've experience with Microsoft.ML.OnnxRuntime library. 
First of all, is it normal that my onnx model is around 600MB? Furthermore I would like to know how to process output from onnx model to obtain the anomaly score; is there a guide or a further documentation to process output feature map from onnx file?

### Describe the solution you'd like

I'd like to get a description of the output structure of onnx model after being executed in C# with Microsoft.ML.OnnxRuntime library. I need to know how can i process the output in order to get the final result as happen making an inference in python.

**Comment by blaz-r** [comment, marked as answer]
Hello. The reason that the model is so big are mean tensor and inverse of covariance matrix inside multivariate Gaussian which can become quite large in dimensions:
https://github.com/openvinotoolkit/anomalib/blob/1f50c952fc65a165884b2b94e178821dcebfbbef/src/anomalib/models/components/stats/multi_variate_gaussian.py#L20-L21
WIth input size of 256x256 they have shape of [100, 4096] and [4096, 100, 100] respectively . With 512x512 they are [100, 16384]  and [16384, 100, 100] respectively. 
Now for pre and post-processing you can refer to [OpenVInoInferencer code](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/deploy/inferencers/openvino_inferencer.py#L30). You will need to handle resizing, normalization and take care of channel order in preprocessing and then normalization and thresholding in post-processing. And for the structure of output, I it should be just a single anomaly map. Again refer to OpenVinoInferencer (and Inferencer base class) for specifics.

**Comment by Tibe31** [comment]
Hi @blaz-r !! 
Thank you for your reply!

**Reply by Dpat123** [reply]
Hi Tibe31

i am also trying to use anomalib from C# . could you please let me know the best method to call training scripts of anomalib from c# and Inference .
have been able to run as per above.

**Reply by samet-akcay** [reply]
Not sure about C# and ONNX, but there is an example here showing C++ and OpenVINO example
https://github.com/openvinotoolkit/model_api/blob/master/model_api/cpp/models/src/anomaly_model.cpp

**Reply by Tibe31** [reply]
Hi @Dpat123,

I didn't call the training scripts from C#. I only run inference using the ONNX model that I obtained after training with python scripts from Anomalib. What I did in C# is the pre-processing on the image and the post-processing -> from the output tensor to the final anomaly-map.

**Reply by Lee-CheWei** [reply]
Hi @Tibe31 ,

I am also trying to use anomalib on C#. Could you please share how to judge the result is "pass" or "fail"?

**Reply by Tibe31** [reply]
Hi @Lee-CheWei,

Since the final output is an 'anomaly map' in which the highest values correspond to
areas of the image for which the model believes there is an anomaly, is up to you to judge if the results is 'pass' or 'fail'.
I suggest, based on your case, to set a threshold where values below it are considered as 'good' and above it as 'anomaly', If all the values of anomaly map are below the threshold you can consider you image as 'pass/good' otherwise 'fail/bad'.

**Comment by Dpat123** [comment]
Ok, thanks for the clarification.

On Fri, Feb 16, 2024 at 10:33 PM Andrea Tiberti ***@***.***>
wrote:

> Hi @Dpat123 <https://github.com/Dpat123>,
>
> I didn't call the training scripts from C#. I only run inference using the
> ONNX model that I obtained after training with python scripts from
> Anomalib. What I did in C# is the pre-processing on the image and the
> post-processing -> from the output tensor to the final anomaly-map.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openvinotoolkit/anomalib/discussions/1539#discussioncomment-8495052>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BFCXQ3HVSJG2MZGECHEZPC3YT6GNRAVCNFSM6AAAAABC7YOWBOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DIOJVGA2TE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
For SquareOne,
Darshan Patel
M-9879565952
